### ISSUE TITLE
[Discussion] Make integrations with asyncio projects more friendly

### ISSUE THREAD

**Issue by standy66** [standy66/purerpc#10]
Right now purerpc uses [curio](https://github.com/dabeaz/curio) event loop, which is really pleasant to use (thx @dabeaz) but limits interoperability with other asyncio projects. Curio was chosen in 2017 mainly because of [this](https://vorpus.org/blog/some-thoughts-on-asynchronous-api-design-in-a-post-asyncawait-world/) @njsmith's post. As of 2019, some things changed:
1. Python 3.7 got decent [upgrade](https://docs.python.org/3/whatsnew/3.7.html#asyncio) to asyncio, some problems mentioned in the blog post were fixed, e.g. StreamWriter now has [wait_closed()](https://docs.python.org/3/library/asyncio-stream.html#asyncio.StreamWriter.wait_closed) method, there is now a [asyncio.get_running_loop()](https://docs.python.org/3/library/asyncio-eventloop.html#asyncio.get_running_loop) function and other new stuff. Some may say, as asyncio evolves and becomes more async/await-native, there is a possibility to migrate to it, especially given that almost all event loop logic in purerpc is abstracted away in [grpc_socket.py](https://github.com/standy66/purerpc/blob/master/src/purerpc/grpc_socket.py). This is one of the paths going forward, and we won't have to think about asyncio interoperability anymore, but there is caveats. Asycnio still lacks curio's wonderful [TaskGroups](https://curio.readthedocs.io/en/latest/reference.html#task-groups), [Thread and Process pools](https://github.com/dabeaz/curio/blob/master/curio/workers.py#L420). 
2. [curio.bridge.AsyncioLoop](https://github.com/dabeaz/curio/blob/master/curio/bridge.py#L25) may be used to bridge together two worlds: asyncio's and curio's. I don't know whether or not there are performance implications and caveats with this approach, but we get to keep awesome curio functionality. We used this bridge internally with [aiopg](https://github.com/aio-libs/aiopg) and it worked fine (cc @penguin138). But we need to make this bridge transparent to the user (maybe some tornado [docs](https://www.tornadoweb.org/en/stable/asyncio.html) may help), so curio and asyncio work together in handlers:
```python3
async some_purerpc_handler(...):
  async for request in requests:
    await asyncio.sleep
    await aiohttp_fetch(request.url)
```
3. We can look at Nathaniel's [trio](https://github.com/python-trio/trio) library which kinda improves on what David has built in curio. There is also trio-asyncio [bridge](https://github.com/python-trio/trio-asyncio), but I've never heard of anyone using it.

**Comment by dabeaz** [comment]
Let me throw out an idea I've been toying around with in my mind....   One of the big ideas in Curio was that of keeping the underlying kernel pretty well isolated from everything else.   I did some experiments with writing a kernel purely in C (which worked).   It might not be too hard to write a Curio kernel that was implemented entirely on top of asyncio.  Maybe it's possible to have some kind of better asyncio interop through that kind of approach.    The idea is that you would have the same high-level Curio programming API that it has now--it'd just be running natively on top of asyncio.  Thoughts?

**Comment by standy66** [comment]
Aren't we gonna get async/await-native API on top of plain old callback API? Surely it's just jumps and system calls from assembler perspective, and OS does not care anyway, but I wonder if there might be any hidden buffers, e.g. for the sake of non-blocking `transport.write`, in the old API that are no longer needed in the new API.

**Comment by njsmith** [comment]
Hey, I hadn't see this before but it's a neat project! Thanks for CC'ing me :-).

@dabeaz 
> It might not be too hard to write a Curio kernel that was implemented entirely on top of asyncio

You'll probably have some issues with `curio.socket`, which exposes a *way* richer API than asyncio does... This is part of why trio-asyncio implements the asyncio on top of trio, instead of vice-versa. In particular, the standard asyncio loop on Windows doesn't even have `wait_writable`/`wait_readable`.

@standy66 
> curio.bridge.AsyncioLoop may be used to bridge together two worlds: asyncio's and curio's. I don't know whether or not there are performance implications and caveats with this approach, but we get to keep awesome curio functionality

One caveat to watch out for is that with `curio.bridge.AsyncioLoop` the curio and asyncio worlds run in actually different threads, so when you have shared state you have to think in terms of thread-programming and possible race conditions everywhere, instead of async-programming and race conditions only at `await`s. This is the main reason why it's useful to get both worlds onto the same underlying loop, whether that's curio-on-top-of-asyncio or asyncio-on-top-of-trio or whatever.

> But we need to make this bridge transparent to the user

We've tried to figure out how to do this in trio-asyncio, but it's really hard :-(. The two worlds are pretty different, especially in their cancellation semantics. Also a lot of libraries (like aiohttp) are really fond of doing things like `asyncio.current_task().cancel()` which is like... totally meaningless in trio, and maybe not quite meaningless in curio, but also not really how curio does things IIRC.

> We can look at Nathaniel's trio library which kinda improves on what David has built in curio

:wave: If you're interested then we'd certainly be happy to help you figure that out (and I think there are probably several people who would be excited about a grpc implementation for trio!). If you have questions then we have a pretty active [chat channel](https://gitter.im/python-trio/general).

Alternatively: You might also want to look at @agronholm's [anyio](https://anyio.readthedocs.io/en/latest/), which provides a mostly-trio-ish API on top of asyncio/curio/trio. There are two major things to be aware of I think, but they probably won't bother you:

1. like curio (but unlike asyncio/trio), anyio uses `await` for all operations, even ones that don't wait for anything; I find this awkward and confusing (esp. because it gets hard to tell which operations are atomic or not), but this may be my own idiosyncracy, and if you're coming from curio you won't notice any difference :-).

2. anyio ignores the high-level networking APIs that asyncio [and trio](https://trio.readthedocs.io/en/latest/reference-io.html#the-abstract-stream-api) provide, in favor of exposing the socket API directly. This means you're stuck dealing with some of the tricky portability issues that these libs would normally handle for you, but curio emphasizes the plain BSD socket API anyway, so this shouldn't be a surprise for you :-). (In particular, anyio has some problems with Windows support because of this.)

Also, using anyio does risk that you might be a bit foreign-feeling everywhere; e.g. Trio has a [standard interface for streaming messages](https://trio.readthedocs.io/en/latest/reference-core.html#using-channels-to-pass-values-between-tasks), and Trio users will probably expect a grpc API to use this for streaming RPCs? But this is probably not the biggest issue on your plate right now...

**Comment by standy66** [comment]
@njsmith thanks for the thorough reply!

I've decided to try @agronholm's anyio library and to my surprise migration went almost seamless. My work is in `anyio` branch and I'll merge #13 once I get CI to run tests for all backends automatically (they can be run manually and already pass though). So I guess you can add PureRPC to the list of libraries that support Trio (and curio and asyncio, of course)!

P.S. I've encountered a bug in anyio's [`sendall` implementation](https://github.com/agronholm/anyio/blob/1e1548d743bb470d2180c453b1a71d07a0f26540/anyio/_networking.py#L145) that causes data corruption. The workaround right now is to [monkeypatch](https://github.com/standy66/purerpc/blob/anyio/src/purerpc/anyio_monkeypatch.py#L21) it to match Dave's implementation in curio. But anyone feel free to send it upstream.

**Comment by agronholm** [comment]
@standy66 I only read this after I fixed the bug. If you are in a hurry for a new release, I can do that. Otherwise I'll first have a go at fixing https://github.com/agronholm/anyio/issues/37.

**Comment by standy66** [comment]
@agronholm Nah, I think I'll be fine if the current workaround stayed for some time. After all, I am already [monkey patching some stuff](https://github.com/standy66/purerpc/blame/anyio/src/purerpc/grpclib/connection.py#L21) in h2 state machine so it behaves itself correctly when communicating with gRPC Core peers.

**Comment by standy66** [comment]
Merged #13 
Going to leave this open for now for further discussion and proposals
